### PR TITLE
Polynomial Detrending Component

### LINKED ISSUE #1944: Polynomial detrending transformer
As an evalml user, I would like to use polynomial detrending in time series regression pipelines.

To close out this issue:
1. Implementation of `PolynomialDetrender` transformer that implements `fit`, `transform`, and `inverse_transform` methods.
2. General solution for applying `inverse_transform` at the pipeline level while predicting/scoring.

Out of scope:

1. Automatic additon to `AutoMLSearch`, i.e. modifying `_get_preprocessing_component`. Reason for this is that I think we can go about it both ways: Always create pipelines with and without detrenders or only add detrenders when we detect a trend. The latter relates to #1730

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Pull Request Description
Fix #1944

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

File: evalml/pipelines/components/component_base_meta.py [hunk 1]
```diff
@@ -19,6 +19,8 @@ def _check_for_fit(self, X=None, y=None):
             klass = type(self).__name__
             if not self._is_fitted and self.needs_fitting:
                 raise ComponentNotYetFittedError(f'This {klass} is not fitted yet. You must fit {klass} before calling {method.__name__}.')
+            elif method.__name__ == 'inverse_transform':
```
Comment (freddyaboulton) [inline evalml/pipelines/components/component_base_meta.py:22]: This is a short-term fix to make the component metaclass aware of the `inverse_transform` method. The current implementation will not work because if `y` is `None`, the method will run with only the `X` argument which is not what we want.

I think we can change this implementation to something like:
```python
def _check_for_fit(self, *args, **kwargs):
    klass = type(self).__name__
    if not self._is_fitted and self.needs_fitting:
        raise ComponentNotYetFittedError(f'This {klass} is not fitted yet. You must fit {klass} before calling {method.__name__}.')
    return method(self, *args, **kwargs)
```

I think this would be more maintainable in the long-term.

File: evalml/pipelines/components/transformers/preprocessing/polynomial_detrender.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            tuple of ww.DataTable, ww.DataColumn: The input features are returned without modification. The target
+                variable y is detrended
+        """
+        if y is None:
+            return X, y
+        y_dt = infer_feature_types(y)
+        y_series = _convert_woodwork_types_wrapper(y_dt.to_series())
+        y_t = self._component_obj.transform(y_series)
+        y_t = ww.DataColumn(pd.Series(y_t, index=y_series.index))
+        return X, y_t
```
Comment (freddyaboulton) [inline evalml/pipelines/components/transformers/preprocessing/polynomial_detrender.py:75]: I'm returning `X` and `y` because our `ComponentGraph` assumes targets are returned as the second value in a tuple

**Comment by freddyaboulton** [comment]
FYI it's impossible to get 100% coverage on the patch:

```python
        # We don't expect to install sktime in python 3.9. Let's verify it's not present:
        if 'sktime' in module and is_running_py_39_or_above:
            with pytest.raises(ModuleNotFoundError):
                import_module(module)
            continue
```

Those lines only run when python 3.9 is installed and we don't measure coverage for python 3.9

File: evalml/tests/component_tests/test_polynomial_detrender.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        PolynomialDetrender(degree=3).fit_transform(X, None)
+
+    pdt = PolynomialDetrender(degree=3).fit(X, y)
+    with pytest.raises(ValueError, match="y cannot be None for PolynomialDetrender!"):
+        pdt.inverse_transform(X, None)
+
+
+@pytest.mark.parametrize("input_type", ['np', 'pd', 'ww'])
+@pytest.mark.parametrize("use_int_index", [True, False])
+@pytest.mark.parametrize("degree", [1, 2, 3])
```
Comment (ParthivNaresh) [inline evalml/tests/component_tests/test_polynomial_detrender.py:31]: Very nice!

**Review by ParthivNaresh — APPROVED** [review]
Everything looks great, nicely done!

**Review by bchen1116 — APPROVED** [review]
Looks great to me! The test coverage is solid.

File: docs/source/install.ipynb [hunk 4]
```diff
@@ -65,7 +65,11 @@
     "\n",
     "```bash\n",
     "brew install graphviz\n",
-    "```"
+    "```\n",
+    "\n",
+    "## Python 3.9 support\n",
+    "\n",
+    "Evalml can still be installed with pip in python 3.9 but note that `sktime`, one of our dependencies, will not be installed because that library does not yet support python 3.9. This means the ``PolynomialDetrending`` component will not be usable in python 3.9. You can try to install `sktime` [from source](https://www.sktime.org/en/latest/installation.html#building-from-source) in python 3.9 to use the ``PolynomialDetrending`` component but be warned that we only test it in python 3.7 and 3.8."
```
Comment (angela97lin) [inline docs/source/install.ipynb:72]: 👍👍

File: evalml/pipelines/components/transformers/preprocessing/polynomial_detrender.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    """Removes trends from time series by fitting a polynomial to the data."""
+    name = "Polynomial Detrender"
+
+    hyperparameter_ranges = {"degree": Integer(1, 3)}
+
+    def __init__(self, degree=1, random_seed=0, **kwargs):
+        """Initialize the PolynomialDetrender.
+
+        Arguments:
+            degree (int): Degree for the polynomial. If 1, linear model is fit to the data.
```
Comment (angela97lin) [inline evalml/pipelines/components/transformers/preprocessing/polynomial_detrender.py:23]: Nitpick: mention default of 1? :)

File: evalml/tests/component_tests/test_polynomial_detrender.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+def test_polynomial_detrender_init():
+    delayed_features = PolynomialDetrender(degree=3)
+    assert delayed_features.parameters == {"degree": 3}
+
+
+def test_polynomial_detrender_raises_value_error_target_is_none(ts_data):
+    X, y = ts_data
+
+    with pytest.raises(ValueError, match="y cannot be None for PolynomialDetrender!"):
```
Comment (angela97lin) [inline evalml/tests/component_tests/test_polynomial_detrender.py:21]: Realllly nit-picking now but do we want to add tests for fit/transform too for completeness :')

File: evalml/tests/component_tests/test_polynomial_detrender.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    expected_answer = pd.Series(detrended_values, index=expected_index)
+
+    X, y = X_input, y_input
+
+    if input_type == 'np':
+        X = X_input.values
+        y = y_input.values
+    elif input_type == 'ww':
+        X = ww.DataTable(X_input)
+        y = ww.DataColumn(y_input)
```
Comment (angela97lin) [inline evalml/tests/component_tests/test_polynomial_detrender.py:55]: We can use the `make_data_type` fixture for this I believe? hehe

**Review by angela97lin — APPROVED** [review]
LGTM, pretty darn cool stuff--the tests were great in understanding what's going on. Left some nitpicky comments per usual but 👍👍👍

File: evalml/tests/component_tests/test_polynomial_detrender.py (hunk 7 above)
Comment (freddyaboulton) [reply evalml/tests/component_tests/test_polynomial_detrender.py:55]: Thanks for the tip! I didn't know about `make_data_type`! I'm not sure if it's worth it in this case because there is no pandas to numpy conversion and the woodwork conversion is straight-forward enough in this case. But I will keep it in mind in the future.

File: evalml/pipelines/components/component_base_meta.py (hunk 1 above)
Comment (freddyaboulton) [reply evalml/pipelines/components/component_base_meta.py:22]: @ParthivNaresh I think we're currently raising an exception in `inverse_transform` if `y` is `None`.  The problem is that the meta class doesn't pass the `y` argument if it's None to `inverse_transform` in the current implementation. This causes an error because we're expecting a `y` argument, even if it's None.

File: evalml/pipelines/components/transformers/preprocessing/polynomial_detrender.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                If 2, quadratic model is fit, etc.
+            random_seed (int): Seed for the random number generator. Defaults to 0.
+        """
+
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/pipelines/components/transformers/preprocessing/polynomial_detrender.py:27]: Great idea!

File: evalml/pipelines/components/transformers/preprocessing/polynomial_detrender.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        super().__init__(parameters=params,
+                         component_obj=detrender,
+                         random_seed=random_seed)
+
+    def fit(self, X, y=None):
+        """Fits the PolynomialDetrender.
+
+        Arguments:
+            X (ww.DataTable, pd.DataFrame, optional): Ignored.
+            y (ww.DataColumn, pd.Series): Target variable to detrend. Ignored.
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/pipelines/components/transformers/preprocessing/polynomial_detrender.py:46]: Typo 😂  Good catch!

File: evalml/tests/component_tests/test_components.py [hunk 10]
```diff
@@ -490,7 +491,8 @@ def all_init(self, *args, **kwargs):
             component = component_class(test_arg="test")
             component_with_different_kwargs = component_class(diff_test_arg="test")
             assert component.parameters['test_arg'] == "test"
-            assert component._component_obj.test_arg == "test"
+            if not isinstance(component, PolynomialDetrender):
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/tests/component_tests/test_components.py:494]: Raises an sklearn exception!

File: requirements.txt [hunk 11]
```diff
@@ -11,3 +11,4 @@ seaborn>=0.11.1
 category_encoders>=2.0.0
 statsmodels >= 0.12.2
 imbalanced-learn>=0.7.0
+sktime>=0.5.3;python_version<"3.9"
```
Comment (dsherry) [inline requirements.txt:14]: 👍  yep good call putting this in requirements and note core-requirements, lgtm

File: evalml/tests/conftest.py [hunk 12]
```diff
@@ -99,6 +100,11 @@ def has_minimal_dependencies(pytestconfig):
     return pytestconfig.getoption("--has-minimal-dependencies")
 
 
+@pytest.fixture
+def is_running_py_39_or_above():
+    return sys.version_info >= (3, 9)
```
Comment (dsherry) [inline evalml/tests/conftest.py:105]: Nifty!

File: evalml/tests/utils_tests/test_dependencies.py [hunk 13]
```diff
@@ -1,17 +1,24 @@
...
     lines = [line for line in lines if '-r ' not in line]
     reqs = requirements.parse(''.join(lines))
     extra_deps = [req.name if req.name != 'imbalanced-learn' else 'imblearn' for req in reqs]
     extra_deps += ['plotly.graph_objects']
     for module in extra_deps:
+        # We don't expect to install sktime in python 3.9. Let's verify it's not present:
+        if module == 'sktime' and is_running_py_39_or_above:
+            with pytest.raises(ModuleNotFoundError):
+                import_module(module)
+            continue
```
Comment (dsherry) [inline evalml/tests/utils_tests/test_dependencies.py:20]: Got it. So the reason this doesn't appear in our coverage report is because we're currently only generating coverage on python 3.8. 👍

File: evalml/pipelines/components/transformers/preprocessing/polynomial_detrender.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        params.update(kwargs)
+        error_msg = "sktime is not installed. Please install using 'pip install sktime'"
+
+        trend = import_or_raise("sktime.forecasting.trend", error_msg=error_msg)
+        detrend = import_or_raise("sktime.transformations.series.detrend", error_msg=error_msg)
+
+        detrender = detrend.Detrender(trend.PolynomialTrendForecaster(degree=degree))
+
+        super().__init__(parameters=params,
+                         component_obj=detrender,
```
Comment (dsherry) [inline evalml/pipelines/components/transformers/preprocessing/polynomial_detrender.py:43]: Is there a reason we need to set `component_obj` for this component? It appears you're doing custom logic for `fit`/`transform`/`fit_transform`, and that was the main utility of passing `component_obj` down to the base class

**Review by dsherry — APPROVED** [review]
🚢 !
